exchanges/graphcache/src/store/data.ts

```diff
@@ -509,27 +515,37 @@ export const reserveLayer = (
   layerKey: number,
   hasNext?: boolean
 ) => {
-  const index = data.optimisticOrder.indexOf(layerKey);
-  if (index === -1) {
-    // The new layer needs to be reserved in front of all other commutative
-    // keys but after all non-commutative keys (which are added by `forceUpdate`)
-    data.optimisticOrder.unshift(layerKey);
-  } else if (!data.commutativeKeys.has(layerKey)) {
-    // Protect optimistic layers from being turned into non-optimistic layers
-    // while preserving optimistic data
-    clearLayer(data, layerKey);
-    // If the layer was an optimistic layer prior to this call, it'll be converted
-    // to a new non-optimistic layer and shifted ahead
-    data.optimisticOrder.splice(index, 1);
-    data.optimisticOrder.unshift(layerKey);
-  }
-
   if (hasNext) {
     data.deferredKeys.add(layerKey);
   } else {
     data.deferredKeys.delete(layerKey);
   }
 
+  let index = data.optimisticOrder.indexOf(layerKey);
+  if (index > -1) {
+    if (hasNext || !data.commutativeKeys.has(layerKey)) {
+      data.optimisticOrder.splice(index, 1);
```

Comment:
It's worth noting that this can cause issues. When a deferred query reorders itself in the layers it may "lift up" old state.
I'd still say this is preferable since overlapping queries are generally discouraged.
However, it'd be cool if we figured out how to implement "sparse layers", i.e. not writing anything in layers that's identical to the lower layers. This may however introduce a significant slowdown to our writes. I haven't measured its impact yet however.

Reply:
Hmm, I wonder if this could be based on selectionSets